src/components/domains/products/ReservationButton.tsx

```diff
@@ -13,40 +13,77 @@ import {
   Text,
   useDisclosure,
 } from '@chakra-ui/react';
+import { ReservationStatus } from '@/api/@types/@enums';
+import { ReservationService } from '@/api/services/Reservation';
+import { RESERVATION_STATUS_COLOR, RESERVATION_STATUS_LABEL } from '@/constants/labels';
 import { useCustomToast } from '@/hooks/useCustomToast';
 
-const ReservationButton = () => {
-  const [isReserved, setIsReserved] = useState<boolean>(false);
+interface ReservationButtonProps {
+  productId: number;
+  reservationStatus: ReservationStatus | undefined;
+  reservationId?: number;
+}
+
+const ReservationButton: FC<ReservationButtonProps> = ({ productId, reservationStatus, reservationId }) => {
   const { isOpen, onOpen, onClose } = useDisclosure();
   const toast = useCustomToast();
   const navigate = useNavigate();
+  const [reservation, setReservation] = useState(() => ({
+    id: reservationId,
+    status: reservationStatus,
+  }));
+
+  const canRequestReservation = !reservation.id;
 
-  const onReserve = () => {
-    setIsReserved(true);
-    toast.success('예약에 성공했어요!');
+  const onReserve = async productId => {
+    try {
+      const { id } = await ReservationService.create(productId);
+      setReservation({
+        id,
+        status: ReservationStatus.REQUESTED,
+      });
+      toast.success('예약 요청에 성공했어요!');
+    } catch (error: any) {
+      if (error.response && error.response.status === 410) {
+        toast.info('이미 예약이 요청되어 있습니다. 판매자 측에서 처리될 때까지 기다려주세요');
+      } else {
+        toast.error('에약 요청에 실패했어요');
+      }
+    }
   };
```

Comment:
```suggestion
  const onReserve = async () => {
    try {
      const { id } = await ReservationService.create(productId);
      setReservation({
        id,
        status: ReservationStatus.REQUESTED,
      });
      toast.success('예약 요청에 성공했어요!');
    } catch (error: any) {
      if (error.response && error.response.status === 410) {
        toast.info('이미 예약이 요청되어 있습니다. 판매자 측에서 처리될 때까지 기다려주세요');
      } else {
        toast.error('에약 요청에 실패했어요');
      }
    }
  };
```

productId 는 동적인 값이 아니기때문에 인자로 받지 않아도 될 것같아요.